src/dawn/Optimizer/PassIntervalPartitioner.cpp

```diff
@@ -0,0 +1,71 @@
+﻿//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#include "dawn/Optimizer/PassIntervalPartitioner.h"
+#include "dawn/IIR/IIRNodeIterator.h"
+#include "dawn/IIR/Stencil.h"
+#include "dawn/IIR/StencilInstantiation.h"
+#include "dawn/Optimizer/OptimizerContext.h"
+
+namespace dawn {
+
+bool PassIntervalPartitioner::run(
+    const std::shared_ptr<iir::StencilInstantiation>& stencilInstantiation) {
+  if(!context_.getOptions().PartitionIntervals) {
+    return true;
+  }
+
+  for(const auto& stencilPtr : stencilInstantiation->getStencils()) {
+    for(auto& multiStage : stencilPtr->getChildren()) {
+      auto cloneMS = multiStage->clone();
+
+      auto multiInterval = multiStage->computePartitionOfIntervals();
+
+      // empty out the multistage
+      for(auto cloneStageIt = cloneMS->childrenBegin(); cloneStageIt != cloneMS->childrenEnd();
+          ++cloneStageIt) {
+        auto cloneDoMethodIt = (*cloneStageIt)->childrenBegin();
+        while(cloneDoMethodIt != (*cloneStageIt)->childrenEnd()) {
+          cloneDoMethodIt = (*cloneStageIt)->childrenErase(cloneDoMethodIt);
```

Comment:
instead of cloning the MS and deleting its data, can't we just create a new multistage?